Feature/quick actions v1

khaykov opened the pull request:
This PR adds the first iteration of static Quick Actions. 

PR looks a bit big, but the majority of changes is from moving things around.

To test:
- Make sure quick action buttons work ok.
- Quick Start, pointers point to the correct rows and blavatar.
- Tracking events are getting hit.
- Nothing looks weird.
- Clicking on Site Name or URL opens site preview.
- Users without access to pages should not see Pages quick action.

![63117921-9b48b380-bf51-11e9-9eb4-14dcb514ade1](https://user-images.githubusercontent.com/728822/63399349-190f2380-c385-11e9-80dc-759bf5eb6505.png)

Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

osullivanchris commented:
Nice work on this. I'm just back from AFK and working through a redline for you @khaykov. I'll have it to you in a couple hours and call out any major differences.

khaykov commented:
@osullivanchris Thanks for the redlines and comments! 

I was using FAB for quick action buttons, so yes, they had a big shadow. I changed it to card shadow that we are using (2dp).
Their dimensions are OS default, so I haven't touched them.

I changed the font type of site title and switch site button to medium and removed bold.

Hope I didn't miss anything!

Also, I was wondering, is there anything extra you maybe want to track?

![Screenshot_1565760536](https://user-images.githubusercontent.com/728822/62996507-ce882700-be19-11e9-9f09-30c28a316a5f.png)

osullivanchris commented:
@khaykov looks much better thanks for making those tweaks!

Just a few more to follow up on:

1. Size of the Quick Action buttons
> Their dimensions are OS default, so I haven't touched them.

Sorry the default dimension for what? A FAB size or something else? I thought these circle buttons were our own design rather than a default component, so I would have gone with the size per the mock. But maybe there's something I'm missing.

2. A few tweaks to the style of site icon
- In the screenshot of the new build, the overall area of the site image with the border seems to be right. But the border looks to be 8dp rather than 4dp. Could you make the border narrower but keep that overall size as is? Correct size is 64dp image, 4dp borders all round making 72dp total (which you have currently just with too much border)
- Rather than a shadow there is a solid line border and no shadow in the mock (value can be seen in the red line)
- Something I didn't call-out in the red line that I now notice is different is that there is a border radius around the image as well as the outer shape. The internal radius is 2dp and the external is 4dp

<img width="419" alt="Screenshot 2019-08-14 at 11 34 12" src="https://user-images.githubusercontent.com/50576199/63014890-8273cb80-be87-11e9-8f38-2e1a7f65a9ca.png">


Might not be exactly how you've written it in code but hope that makes sense! Thanks!

khaykov commented:
> Sorry the default dimension for what? A FAB size or something else? I thought these circle buttons were our own design rather than a default component, so I would have gone with the size per the mock. But maybe there's something I'm missing.

The dimension of FAB. Quick action buttons are just FAB's, so they already have correct dimensions.

> In the screenshot of the new build, the overall area of the site image with the border seems to be right. But the border looks to be 8dp rather than 4dp. Could you make the border narrower but keep that overall size as is? Correct size is 64dp image, 4dp borders all round making 72dp total (which you have currently just with too much border)

Good catch! Totally missed that one.

> Rather than a shadow there is a solid line border and no shadow in the mock (value can be seen in the red line)

👍 

> Something I didn't call-out in the red line that I now notice is different is that there is a border radius around the image as well as the outer shape. The internal radius is 2dp and the external is 4dp

👍 

I think it looks about right now :)
![Screenshot_1565824213](https://user-images.githubusercontent.com/728822/63062498-0fca1680-beae-11e9-9570-aa038fa336a4.png)

osullivanchris commented:
@khaykov Yup looking great now. Checked with @SylvesterWilmott and yes the Quick Actions should be the same size as FABs. When I'm measuring it here it looks like the buttons are 48dp when I would expect 56. But it might be that I'm looking at the wrong density. Per the guidelines the only sizes are 56 and 40. Assuming you're using 56 already as 40 would be much too small.
<img width="798" alt="Screenshot 2019-08-15 at 14 37 29" src="https://user-images.githubusercontent.com/50576199/63098210-46656700-bf6a-11e9-85ee-5679888eaf9c.png">

The guidelines also mention only having one FAB per page. As long as it looks right I'm ok. And it looks right (just double checking that detail above). Just wanted to mention this in case there is any semantic issues or incase we get dinged in app store review or elsewhere. As I said once it looks like the mock I'm happy!
<img width="392" alt="Screenshot 2019-08-15 at 14 38 52" src="https://user-images.githubusercontent.com/50576199/63098260-6bf27080-bf6a-11e9-9dc9-2bb3fc71c844.png">


Two other small details that came up chatting to Sylvester
- Can we make the quick start card the same style as the stats card (full-width and no shadow). It was designed to match the card above which we have now killed, so it looks a bit off having one thing as a card. We can make a separate task for that if you want
- Can you add the white background to the 'Register domain' row so it matches the other rows below it?

khaykov commented:
@osullivanchris Thanks for the guidance! I updated layout:

![ab](https://user-images.githubusercontent.com/728822/63117921-9b48b380-bf51-11e9-9eb4-14dcb514ade1.png)


> The guidelines also mention only having one FAB per page. As long as it looks right I'm ok. And it looks right (just double checking that detail above). Just wanted to mention this in case there is any semantic issues or incase we get dinged in app store review or elsewhere. As I said once it looks like the mock I'm happy!

I double-checked and the size is correct 👍Also, it's FAB only in code, for all other purposes it's just a round button.

osullivanchris commented:
All looks great @khaykov thanks for going through those extra details!

theck13 commented on WordPress/src/main/res/layout/my_site_fragment.xml:
```diff
@@ -2,634 +2,677 @@
+                        <com.google.android.material.floatingactionbutton.FloatingActionButton
+                            android:id="@+id/quick_action_pages_button"
+                            style="@style/MySiteQuickActionButton"
+                            android:backgroundTint="@android:color/white"
```
```suggestion
```

We can remove the `android:backgroundTint="@android:color/white"` attribute from the `quick_action_pages_button` view since the `MySiteQuickActionButton` style already has that attribute.  That would eliminate a redundancy and make the view consistent with the other Quick Action buttons.

theck13 commented on WordPress/src/main/res/layout/my_site_fragment.xml:
```diff
@@ -2,634 +2,677 @@
                     android:layout_height="wrap_content"
-                    android:layout_width="match_parent" >
+                    android:layout_marginBottom="@dimen/margin_large"
+                    android:background="@color/white"
```
```suggestion
                    android:background="@android:color/white"
```

Can we change this to `@android:color/white`?  There is no resource `white` in `colors.xml` so using `@color/white` would be pulling from another source or library.

theck13 commented on WordPress/src/main/res/layout/my_site_fragment.xml:
```diff
@@ -2,634 +2,677 @@
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_marginBottom="@dimen/margin_large"
+                    android:background="@color/white"
```
```suggestion
                    android:background="@android:color/white"
```

Can we change this to `@android:color/white`?  There is no resource `white` in `colors.xml` so using `@color/white` would be pulling from another source or library.

theck13 commented on WordPress/src/main/res/values/dimens.xml:
```diff
@@ -81,11 +81,11 @@
     <dimen name="margin_large">12dp</dimen>
     <dimen name="margin_extra_large">16dp</dimen>
     <dimen name="margin_extra_medium_large">24dp</dimen>
-    <dimen name="margin_extra_extra_medium_large">32dp</dimen>
```
This dimension might have been removed accidentally.  It looks like it's still used [here](https://github.com/wordpress-mobile/WordPress-Android/pull/10362/files?file-filters%5B%5D=.java&file-filters%5B%5D=.xml#diff-7c2f033014a455e52db8dd994c5f3e58R35) and [here](https://github.com/wordpress-mobile/WordPress-Android/pull/10362/files?file-filters%5B%5D=.java&file-filters%5B%5D=.xml#diff-7c2f033014a455e52db8dd994c5f3e58R116).  The same dimension is defined in `WordPressLoginFlow` so those instances are using that reference.  Should we add it back to `WordPress` so that it is explicitly defined in the main module too?

khaykov replied:
Not sure how it even got removed. Mystery.

theck13 commented on WordPress/src/main/res/values/styles.xml:
```diff
@@ -269,12 +269,38 @@
         <item name="android:paddingEnd">@dimen/content_margin_site_row_end</item>
         <item name="android:paddingStart">@dimen/content_margin_site_row_start</item>
-        <item name="android:background">?android:attr/selectableItemBackground</item>
+        <item name="android:background">@color/white</item>
```
```suggestion
        <item name="android:background">@android:color/white</item>
```

Can we change this to `@android:color/white`?  There is no resource `white` in `colors.xml` so using `@color/white` would be pulling from another source or library.

khaykov commented:
Thanks for a thorough review, @theck13 !

> The touch feedback for the Switch Site button isn't what I would expect from a text button. Material guidelines show a rectangular background when pressing text buttons. Should we change the background from ?attr/selectableItemBackgroundBorderless to ?attr/selectableItemBackground and mimic the background border of the Change Photo text button on the Me tab? See the screenshots below for illustration.

Good catch! I changed the background and added a little bit of padding to the sides so it will look better.

> The Quick Action buttons look well-aligned with the interface below them on a typical phone form factor. On a tablet, they extend beyond the text. Should we use the content margin dimensions to keep the Quick Action buttons aligned with the interface below them on large sccreen? See the screenshots below for illustration.

I agree, but honestly not sure how to approach this. We are using layout_weight and evenly spreading button, so any margin will mess this up. I'm open to ideas :) 

I added content margin to the site title and address tho :)

![Screenshot_1566444490](https://user-images.githubusercontent.com/728822/63484117-72458880-c453-11e9-8e13-5c8d2a3f1372.png)

khaykov commented on WordPress/src/main/res/layout/my_site_fragment.xml:
```diff
@@ -2,634 +2,677 @@
 
-                    </RelativeLayout>
+                <LinearLayout
+                    android:id="@+id/site_info_container"
```
Fixed it! The clickable area spans the whole screen (hope it's ok!).

osullivanchris commented:
Great details to spot, thanks @theck13 

Agree with all of those. Looks like the button touch area, and the content padding have already been adjusted.

> I agree, but honestly not sure how to approach this. We are using layout_weight and evenly spreading button, so any margin will mess this up. I'm open to ideas :)

Is it not possible to spread the buttons evenly within an area? Not sure how it would work in code. But could it be a separate div with margin outside of it, and the items spaced equally within it? Alternatively I could provide paddings between the buttons, but not sure that will be as successful across device sizes as evenly spacing them

khaykov commented:
@osullivanchris Hm. Maybe I'm missing something? 

Currently, buttons are spread evenly, like this - you can see four even boxes, and buttons float in the center of each box.
[![Image from Gyazo](https://i.gyazo.com/6df714ae668f6bdb455ca553803aaf49.png)](https://gyazo.com/6df714ae668f6bdb455ca553803aaf49)

If we add padding or margin, the container with 4 boxes will get smaller, boxes will get smaller, but the buttons will still float in the center (will be spread evenly withing smaller container).
[![Image from Gyazo](https://i.gyazo.com/abd6f6a383376b849949492707fe19d7.png)](https://gyazo.com/abd6f6a383376b849949492707fe19d7)

After adding padding, we can tell first and last button to stick to sides, instead of floating in the center of their box, this way we will get them to be extacly where we want, but this will make other buttons look weird:
[![Image from Gyazo](https://i.gyazo.com/1fd621e4ac96ecf6429110900d112efb.png)](https://gyazo.com/1fd621e4ac96ecf6429110900d112efb)

We can also make two buttons in the center gravitate towards their side of the screen. Looks more symmetrical, but we have a weird gap.
[![Image from Gyazo](https://i.gyazo.com/10bfe53de67816c3da066f6dc2a0a493.png)](https://gyazo.com/10bfe53de67816c3da066f6dc2a0a493)

And there is the case with 3 buttons, that some users will see
[![Image from Gyazo](https://i.gyazo.com/24d451dd9b67262f14f268e6d9eb8c1c.png)](https://gyazo.com/24d451dd9b67262f14f268e6d9eb8c1c)

osullivanchris commented:
> If we add padding or margin, the container with 4 boxes will get smaller, boxes will get smaller, but the buttons will still float in the center (will be spread evenly withing smaller container).

this sounds like what I would expect 

I'm picturing it that as the screen width increases there becomes a larger margin area (in blue) where content will not spread
<img width="719" alt="Screenshot 2019-08-22 at 18 12 09" src="https://user-images.githubusercontent.com/50576199/63535229-dfffbc00-c508-11e9-9aec-ac3d31ed07a8.png">

If as the screen width increases and the buttons are spread equally, the outside button doesn't line up with the left edge of the content below it, thats fine by me. Example here 
<img width="1125" alt="Screenshot 2019-08-22 at 18 12 17" src="https://user-images.githubusercontent.com/50576199/63535260-f6a61300-c508-11e9-83bc-0136d64f5d2d.png">

I don't think the outside of the buttons need to be strictly aligned to the edge of the list content. I would rather they end up narrower than the list, rather than spreading out wider than it. If this is confusing I can explain better on a call tomorrow or something! :)

khaykov commented:
@osullivanchris Got it! For some reason, I thought that buttons should align with rest of the text :) 
I added a content margin, and it looks like this on tablet:
[![Image from Gyazo](https://i.gyazo.com/9ed5e1a95a2bb6c9bc3bb4571dda96ce.png)](https://gyazo.com/9ed5e1a95a2bb6c9bc3bb4571dda96ce)

And like this on the phone:
[![Image from Gyazo](https://i.gyazo.com/025ecaf53dbbfd0080d42571b5d1c7f3.png)](https://gyazo.com/025ecaf53dbbfd0080d42571b5d1c7f3)

theck13 commented on WordPress/src/main/res/layout/my_site_fragment.xml:
```diff
@@ -2,634 +2,677 @@
+                    android:fontFamily="sans-serif-medium"
+                    android:gravity="center_vertical"
+                    android:maxLines="1"
+                    android:text="@string/my_site_btn_switch_site"
```
```suggestion
                    android:text="@string/my_site_btn_switch_site"
                    android:padding="@dimen/margin_medium"
```

Can we add the `android:padding="@dimen/margin_medium"` attribute to the `switch_site` view?  That will add left/start and right/end padding to the text button with the default text size as well as ensure the top and bottom padding exists for large text sizes.

khaykov replied:
Huh. I thought I have added padding already 😕

theck13 commented:
@osullivanchris, I see you approved these changes a week ago, but I wanted to confirm you approve the changes of the last two days as well.  Are we good to merge this?

osullivanchris commented:
@khaykov ah cool, glad we were able to clear it up. Looks good!

@theck13 yep I'm happy with it, thanks for checking
